### PR TITLE
Minor client/server tunnel refactor

### PR DESCRIPTION
**Pull request by MoroccanMalinois**
---
**By submitting this pull-request, I confirm the following:**

- I have read and understood the [Contributing Guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that another pull-request for this purpose does not exist.
- I have considered and confirmed that this submission will be valuable to others.
- I accept that this submission may not be used and that this pull-request may be closed by the will of the maintainer.
- I give this submission freely under the BSD 3-clause license.
---
Minor changes, just preparation for a future discussion on IRC.
Referencing #96

### CODE REVIEW

File: src/client/context.cc [hunk 1]
```diff
@@ -385,6 +371,50 @@ void ClientContext::UpdateClientTunnel(
...
+  // Insert client tunnel
+  auto client_tunnel = std::make_unique<kovri::client::I2PClientTunnel>(
+      tunnel, local_destination);
+  if (!InsertClientTunnel(tunnel.port, std::move(client_tunnel)))
+    {
+      LOG(error) << "Instance: failed to insert new client tunnel";
+      return false;
+    }
+  std::lock_guard<std::mutex> lock(m_ClientMutex);
+  m_ClientTunnels[tunnel.port]->Start();
```
Comment (MoroccanMalinois) [inline src/client/context.cc]: This is the only logic change. ClientContext::AddClientTunnel is called from 2 places : one of them didn't have the Start call.  I don't know if it should be called, but i found it weird that they didn't do the same.

File: src/app/instance.cc [hunk 2]
```diff
@@ -212,17 +186,67 @@ void Instance::SetupTunnels() {
...
+              kovri::client::context.UpdateServerTunnel(tunnel, is_http);
+              ++server_count;
+            }
+        }
+      catch (const std::exception& ex)
+        {
+          LOG(error) << "Instance: exception during tunnel setup: "
+                     << ex.what();
+          return;
+        }
```
Comment (anonimal) [inline src/app/instance.cc]: You can get rid of this block and implement the exception dispatcher in the next block.

**Comment by MoroccanMalinois** [comment]
@anonimal : it's little better, still multiple todo's in this area but i think it's good enough for a first review.

File: src/app/instance.h [hunk 3]
```diff
@@ -81,9 +81,7 @@ class Instance {
   void SetupTunnels(bool is_reloading = false);
 
   /// @brief Should remove old tunnels after tunnels config is updated
-  /// TODO(unassigned): not fully implemented
-  void RemoveOldTunnels(
-      std::vector<std::string>& updated_tunnels);
+  void RemoveOldTunnels(std::vector<std::string>& updated_tunnels);
```
Comment (anonimal) [inline src/app/instance.h]: The commit says `Instance: fix RemoveOldTunnels` but `RemoveOldTunnels()` is still not fully implemented (see definition). Can you please keep the TODO (or fully implement)?

File: src/app/instance.cc [hunk 4]
```diff
@@ -80,12 +78,11 @@ void Instance::Initialize() {
 
 void Instance::Reload() {
   LOG(info) << "Instance: reloading";
-  // TODO(unassigned): reload kovri.conf
   // TODO(unassigned): locking etc.
   // TODO(unassigned): client/router contexts
-  m_IsReloading = true;
-  SetupTunnels();
-  m_IsReloading = false;
+  GetConfig().ParseKovriConfig();
```
Comment (anonimal) [inline src/app/instance.cc]: This doesn't work (`kovri.conf` re-loading). It won't work without work elsewhere (see other TODO's). Can you please remove this line or implement?

File: src/client/context.cc [hunk 5]
```diff
@@ -370,8 +356,9 @@ void ClientContext::UpdateClientTunnel(
     if (ec)  // New address is not an IP address, compare strings
       rebind = (tunnel.address != current_addr);
     else  // New address is an IP address, compare endpoints
-      rebind = (client_tunnel->GetEndpoint() == boost::asio::ip::tcp::endpoint(
-          next_addr, tunnel.port));
+      rebind =
+          (client_tunnel->GetEndpoint()
+           != boost::asio::ip::tcp::endpoint(next_addr, tunnel.port));
```
Comment (anonimal) [inline src/client/context.cc]: 👍

File: src/app/instance.h [hunk 6]
```diff
@@ -76,23 +76,17 @@ class Instance {
   void InitClientContext();
 
   /// @brief Sets up (or reloads) client/server tunnels
+  /// @param is_reloading Are tunnels configuration in the process of reloading?
   /// @warning Configuration files must be parsed prior to setup
-  void SetupTunnels();
+  void SetupTunnels(bool is_reloading = false);
```
Comment (anonimal) [inline src/app/instance.h]: >removing member m_IsReloading

1. Asumming that locks are not where they should be, if `Instance::Reload()` is called consecutively (multiple consecutive `SIGHUP`'s), I think we'll be looking at possible race conditions

2. `SetupTunnels()` with member `m_IsReloading` is a cheap way to resolve point 1 above

3. I find bool `is_reloading` as function arg to be somewhat confusing but I think the real issue is *how* we setup/reload tunnels.

Any other thoughts? How else could we approach these issues without more patching?

File: src/client/context.h [hunk 7]
```diff
@@ -174,20 +186,16 @@ class ClientContext {
...
 
   /// @return the client tunnel with the given name, or nullptr
-  std::unique_ptr<I2PClientTunnel> GetClientTunnel(
-      const std::string& name);
+  I2PClientTunnel* GetClientTunnel(const std::string& name);
 
   /// @return the client tunnel with the given (local) port
-  std::unique_ptr<I2PClientTunnel> GetClientTunnel(
-      int port);
+  I2PClientTunnel* GetClientTunnel(int port);
```
Comment (anonimal) [inline src/client/context.h]: >raw pointer refactoring

I'm glad my suggestions on IRC worked out 😄

File: src/client/context.h [hunk 8]
```diff
@@ -138,12 +138,24 @@ class ClientContext {
...
+  /// @param http true if server tunnel is an HTTP tunnel
+  /// @return true if the tunnel was inserted, false otherwise
+  bool AddServerTunnel(
+      const TunnelAttributes& tunnel,
+      bool is_http);  // TODO(anonimal): this isn't ideal
+
+  /// @brief Creates the specified client tunnel and tries to insert it
+  /// @param tunnel Const reference to populated/initialized tunnel attributes class
+  /// @return true if the tunnel was inserted, false otherwise
+  bool AddClientTunnel(const TunnelAttributes& tunnel);
```
Comment (anonimal) [inline src/client/context.h]: >pulling out tunnel creation functionality
👍

File: src/app/instance.h (hunk 3 above)
Comment (MoroccanMalinois) [inline src/app/instance.h]: Sorry @anonimal , but i really don't see what's left to do. This function removes all tunnels that are not present in "updated_tunnels". What am i missing ?

Comment (anonimal) [inline src/app/instance.h]: AFAICT, nothing! I was distracted and overlooked the 2 important lines that were already implemented. My sincerest apologies.

File: src/app/instance.h (hunk 6 above)
Comment (MoroccanMalinois) [inline src/app/instance.h]: I removed this commit. I guess this falls in "// TODO(unassigned): locking etc."

File: src/app/instance.cc (hunk 4 above)
Comment (MoroccanMalinois) [inline src/app/instance.cc]: Line removed

**Pull request merged**